### PR TITLE
updates to conduit and ascent packages

### PR DESCRIPTION
**Pull request by cyrush**
Updates to conduit and ascent packages to:
* provide an optional path to use spack's py-site-pkgs dir
* install generated host-cfg for provenance

### CODE REVIEW

**Comment by cyrush** [comment]
FYI: This is an important change to avoid issues in our automated build process related to #6261.

Also, even if #6261 was resolved, this is still a cleaner solution -- b/c it will allow my subclass to directly dictate the site packages dir. (It's not clear the base site-packages dir would be correct)

File: var/spack/repos/builtin/packages/conduit/package.py [hunk 1]
```diff
@@ -209,9 +213,9 @@ def create_host_config(self, spec, prefix):
                 raise RuntimeError(msg)
             cmake_exe = cmake_exe.path
 
-        host_cfg_fname = "%s-%s-%s.cmake" % (socket.gethostname(),
-                                             sys_type,
-                                             spec.compiler)
+        host_cfg_fname = "%s-%s-%s-conduit.cmake" % (socket.gethostname(),
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/conduit/package.py]: Was this rename required? Or was it just cosmetic?

File: var/spack/repos/builtin/packages/ascent/package.py [hunk 2]
```diff
@@ -246,10 +249,10 @@ def create_host_config(self, spec, prefix):
             cfg.write("# python from spack \n")
             cfg.write(cmake_cache_entry("PYTHON_EXECUTABLE",
                       spec['python'].command.path))
-            # install module to standard style site packages dir
-            # so we can support spack activate
-            cfg.write(cmake_cache_entry("PYTHON_MODULE_INSTALL_PREFIX",
-                                        site_packages_dir))
+            # only set dest python site packages dir if passed
+            if py_site_pkgs_dir is not None:
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/ascent/package.py]: Would you be using `py_site_pkgs_dir` if it were `''`? In other words, if `py_site_pkgs_dir` would always be a nonempty string when you intend to use it, you can say `if py_site_pkgs_dir:`.

File: var/spack/repos/builtin/packages/ascent/package.py [hunk 3]
```diff
@@ -140,9 +142,10 @@ def install(self, spec, prefix):
             cmake(*cmake_args)
             make()
             make("install")
-            # TODO also copy host_cfg_fname into install
+            # install copy of host config for provenance
+            install(host_cfg_fname, prefix)
 
-    def create_host_config(self, spec, prefix):
+    def create_host_config(self, spec, prefix, py_site_pkgs_dir=None):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/ascent/package.py]: Could you add a comment which refers to https://github.com/spack/spack/issues/6261 and/or explain that this ensures that the site-packages directory is defined when this is called from a subclass?

As a side note, when you mention:

> It's not clear the base site-packages dir would be correct

I'm not sure what you mean. The spec you would be working with here is the subclass spec, so you could construct the subclass-specific site packages directory. That being said the way it's done here seems straightforward (if the comment is added).

File: var/spack/repos/builtin/packages/conduit/package.py (hunk 1 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/conduit/package.py]: w/o the rename, I can't spack activate those packages at the same time (files with the same name conflict)

File: var/spack/repos/builtin/packages/ascent/package.py (hunk 2 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/ascent/package.py]: Sure thing, happy to do this. 

Sorry this my general python style is leaking though.

File: var/spack/repos/builtin/packages/ascent/package.py (hunk 3 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/ascent/package.py]: sure I can add a comment about #6261 

The "it's not clear" is b/c the `site_package_dir` is set as a module var, not as part of the package instance.  But it seems to be set to specific install location of the package instance. 

Even finding the right path, what would the base class `site_packages_dir` be set to?
The right answer is not clear to me. The mix of inheritance + magic makes it a bit hard for me the comprehend. But thats all fine, I think the solution I have is good b/c it allows me to still use spack's site_packages_dir as intended for the main packages.

**Comment by cyrush** [comment]
@scheibelp  thanks for the feedback, I'll try to resolve these issues today!

**Comment by cyrush** [comment]
@scheibelp pending CI-checks, but should be ready

File: var/spack/repos/builtin/packages/conduit/package.py [hunk 4]
```diff
@@ -163,14 +165,25 @@ def install(self, spec, prefix):
             cmake(*cmake_args)
             make()
             make("install")
+            # install copy of host config for provenance
+            install(host_cfg_fname, prefix)
 
-    def create_host_config(self, spec, prefix):
+    def create_host_config(self, spec, prefix, py_site_pkgs_dir=None):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/conduit/package.py]: @cyrush Sorry one last question: when is `py_site_pkgs_dir` not set by the caller? When called from `Conduit.install` it is set, and I assumed the subclass would also want to set it too.

(this question also applies to `ascent` since the same approach was taken there.

Comment (cyrush) [reply var/spack/repos/builtin/packages/conduit/package.py]: When the python variant is off, I don't want the call to require `py_site_pkgs_dir`.

I suspect I could pass `""`, but I usually use None for optional args. 

Thinking about this, I actually need to test `~python`, I don't think it's wired up correctly in this PR

It was guarded by if `+python` in spec in the previous iteration, I lost that with the new call.

I'll pulse when that is fixed + tested.

**Comment by cyrush** [comment]
@scheibelp

fixed to no longer call `site_packages_dir` when  `~python`, also tested builds for +doc and found and resolved and issue with ascent package.py.

Let me know if there is anything else that needs to be resolved.

**Comment by cyrush** [comment]
I have one more fix for these packages : capturing proper mpiexec - or equivalent.  I'll be working on this today

**Comment by cyrush** [comment]
closed and reopened to re-run CI checks. Failures didn't seem related to my changes.

**Comment by cyrush** [comment]
@scheibelp this is ready to go, pending understanding of the code coverage issues

**Comment by cyrush** [comment]
tried to rebase, and apparently messed up b/c its including all changes since i forked.
